{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# To recap\n",
    "\n",
    "### Data so far\n",
    "\n",
    "1. We have created a list of people and their ORCID identifiers - [output/pure_person_to_orcid.txt](output/pure_person_to_orcid.txt)\n",
    "2. We have created a list of ORCID to PubMed identifiers - [output/orcid.tsv](output/orcid.tsv)\n",
    "3. We have created a list of PubMed IDs to PubMed info - [output/pubmed.tsv](output/pubmed.tsv)\n",
    "4. We have created a list of the top 100 terms for each person - [output/orcid-tf-idf.txt](output/orcid-tf-idf.txt)\n",
    "\n",
    "### Questions\n",
    "\n",
    "Now we need to use the common identifiers in these data to answer some questions:\n",
    "\n",
    "1. Can we produce a set of potential collaborators for each person, a collaborator being someone they have significant terms in common with but not previously published with.\n",
    "2. Can we select a set of people that most closely map to a specific piece of text.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pandas dataframes\n",
    "\n",
    "Normally at this point we would use a database. However, to keep it all in a single language and framework, and as the data themselves are relatively small, we can use pandas - https://pandas.pydata.org/\n",
    "\n",
    ">pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, following on from the data descriptions above:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 1. People and their ORCID identifiers \n",
    "personToOrcid = pd.read_csv('data/pure_person_to_orcid.txt',sep='\\t')\n",
    "print(personToOrcid.shape)\n",
    "print(personToOrcid.head())\n",
    "\n",
    "# 2. ORCID to PubMed identifiers\n",
    "orcidToPubmed = pd.read_csv('data/orcid.tsv',sep='\\t')\n",
    "print(orcidToPubmed.shape)\n",
    "print(orcidToPubmed.head())\n",
    "\n",
    "# 3. PubMed IDs to PubMed info\n",
    "pubmedToInfo = pd.read_csv('data/pubmed.tsv',sep='\\t')\n",
    "print(pubmedToInfo.shape)\n",
    "print(pubmedToInfo.head())\n",
    "\n",
    "# 4. Top 100 terms for each person\n",
    "topTerms = pd.read_csv('data/orcid-tf-idf.txt',sep='\\t')\n",
    "print(topTerms.shape)\n",
    "print(topTerms.head())\n",
    "\n",
    "# 5. PURE to name\n",
    "pureToName = pd.read_csv('data/pure_people.txt',sep='\\t')\n",
    "print(pureToName.shape)\n",
    "print(pureToName.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A first pass would be to count the number of matching terms between each person. To start with, let's just compare the first in the list to everyone else:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " #loop through people (limit to top 1)\n",
    "def compare_people(limit):\n",
    "    for i, iData in personToOrcid.head(n=limit).iterrows():\n",
    "        print('### Comparing',i,iData.pure_person_id,iData.orcid_id)\n",
    "        #get all terms and tf-idf values\n",
    "        iTopTerms=topTerms[topTerms['orcid_id']==iData.orcid_id]\n",
    "        #print(iTopTerms.head())\n",
    "        iTerms = iTopTerms['term']\n",
    "\n",
    "        for j in range(i+1,personToOrcid.shape[0]):\n",
    "            jData=personToOrcid.iloc[j]\n",
    "            jTopTerms=topTerms[topTerms['orcid_id']==jData.orcid_id]\n",
    "            #print(jTopTerms['term'].head())\n",
    "            jTerms = jTopTerms['term']\n",
    "            #print(iTerms,jTerms)\n",
    "            com = list(set(iTerms) & set(jTerms))\n",
    "            #only show maches with more than 1 matching term\n",
    "            if len(com) > 1:\n",
    "                print(jData.orcid_id,com)\n",
    "    \n",
    "compare_people(1)        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are few matches with many overlapping terms, except one **0000-0001-7086-8882**. Let's look at the research pages of these two people\n",
    "\n",
    "https://research-information.bristol.ac.uk/en/persons/melody-a-s-sylvestre(81e7f06c-77d8-4020-9608-0b30dd001c43)/publications.html https://research-information.bristol.ac.uk/en/persons/nicholas-a-teanby(4ec18a96-fd2e-4311-a6fa-7ec65696a4e9)/publications.html \n",
    "\n",
    "This suggests they are from similar research areas, and have indeed co-published.   \n",
    "\n",
    "Let's try again, this time with the top 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "compare_people(3)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second person matched no-one, the third however matched quite a few. Let's modify the function to sort by number:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_people(limit):\n",
    "    for i, iData in personToOrcid.head(n=limit).iterrows():\n",
    "        print('### Comparing',i,iData.pure_person_id,iData.orcid_id)\n",
    "        #get all terms and tf-idf values\n",
    "        iTopTerms=topTerms[topTerms['orcid_id']==iData.orcid_id]\n",
    "        iTerms = iTopTerms['term']\n",
    "        jComp={}\n",
    "        for j in range(i+1,personToOrcid.shape[0]):\n",
    "            jData=personToOrcid.iloc[j]\n",
    "            jTopTerms=topTerms[topTerms['orcid_id']==jData.orcid_id]\n",
    "            jTerms = jTopTerms['term']\n",
    "            com = list(set(iTerms) & set(jTerms))\n",
    "            #only show maches with more than 1 matching term\n",
    "            if len(com) > 1:\n",
    "                jComp[jData.orcid_id]=com\n",
    "        #create sorted dictionary using number of items\n",
    "        jComp = sorted(jComp.items(), key=lambda kv: len(kv[1]), reverse=True)\n",
    "        for p in jComp[0:10]:\n",
    "            print(p)\n",
    "            \n",
    "compare_people(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This only shows us terms that are in common. It could be possible to have lots of terms in common, but these terms are not the 'most' representative of either person. It would be better to compare all terms, and their tf-idf values.\n",
    "\n",
    "Back in the TF-IDF section (http://localhost:8888/lab#TF-IDF-using-sklearn) we created a similarity matrix from the TF-IDF model.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get similarity matrix for all people\n",
    "\n",
    "matrixCom = {}\n",
    "\n",
    "#get tfidf matrix and orcidText dictionary\n",
    "%store -r matrix\n",
    "%store -r token_dict\n",
    "\n",
    "for i in range(0,len(matrix)):\n",
    "    iOrcid=list(token_dict)[i]\n",
    "    matrixCom[iOrcid]={}\n",
    "    for j in range(0,len(matrix)):\n",
    "        jOrcid=list(token_dict)[j]\n",
    "        matrixCom[iOrcid][jOrcid]=matrix[i][j]\n",
    "\n",
    "o = open('output/orcid-to-orcid-tf-idf.tsv','w')\n",
    "o.write('orcid_1\\torcid_2\\ttf-idf\\n')\n",
    "counter=0\n",
    "for m in matrixCom:\n",
    "    sorted_res = sorted(matrixCom[m].items(), key=lambda kv: kv[1], reverse=True)\n",
    "    if counter<3:\n",
    "        print(m)\n",
    "        for s in sorted_res[0:5]:\n",
    "            print('\\t',s[0],s[1])\n",
    "        counter+=1\n",
    "    for s in sorted_res:\n",
    "        o.write(m+'\\t'+s[0]+'\\t'+str(s[1])+'\\n')\n",
    "o.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Collaboration recommendation engine\n",
    "\n",
    "This similarity data depicts the similarity between each person's publication text based on tf-idf. Often, similarities arise due to co-publication, and perhaps a more informative recommender would be to idenfity cases where people have signficant overlap in their publication text, but have never previously co-published. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load the new data into a dataframe\n",
    "\n",
    "orcidToOrcid = pd.read_csv('data/orcid-to-orcid-tf-idf.tsv',sep='\\t')\n",
    "print(orcidToOrcid.shape)\n",
    "print(orcidToOrcid.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0000-0002-1407-8314 536    0.102831\n",
      "       term  tf-idf_x  tf-idf_y\n",
      "0    cancer  0.085914  0.079305\n",
      "1     genes  0.061123  0.050338\n",
      "2      data  0.054236  0.064645\n",
      "3     using  0.035485  0.060613\n",
      "4  analysis  0.032270  0.051746 \n",
      "\n",
      "0000-0003-1562-891X 537    0.098246\n",
      "            term  tf-idf_x  tf-idf_y\n",
      "0           cell  0.091965  0.151794\n",
      "1         cancer  0.085914  0.264330\n",
      "2         breast  0.059255  0.247853\n",
      "3  breast cancer  0.038263  0.198393\n",
      "4    cancer cell  0.032882  0.042547\n",
      "5    development  0.031938  0.032471 \n",
      "\n",
      "0000-0001-8186-5708 542    0.074373\n",
      "     term  tf-idf_x  tf-idf_y\n",
      "0    cell  0.091965  0.274139\n",
      "1  cancer  0.085914  0.062854 \n",
      "\n",
      "0000-0002-2715-9930 543    0.071174\n",
      "          term  tf-idf_x  tf-idf_y\n",
      "0       cancer  0.085914  0.042130\n",
      "1        genes  0.061123  0.115058\n",
      "2         data  0.054236  0.045467\n",
      "3   mechanisms  0.040592  0.030163\n",
      "4  development  0.031938  0.035598 \n",
      "\n",
      "0000-0002-6009-7137 544    0.070957\n",
      "     term  tf-idf_x  tf-idf_y\n",
      "0    cell  0.091965  0.241663\n",
      "1  cancer  0.085914  0.108549\n",
      "2    stem  0.070103  0.028470\n",
      "3   tumor  0.038173  0.032943 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#filter on co-publishing\n",
    "example_orcid='0000-0001-7328-4233'\n",
    "#example_orcid='0000-0003-0924-3247'\n",
    "example_pubs=orcidToPubmed[orcidToPubmed['orcid_id']==example_orcid]['pmid']\n",
    "\n",
    "def find_overlapping_tfidf(orcid1,orcid2):\n",
    "    o1=topTerms[topTerms['orcid_id']==orcid1]\n",
    "    o2=topTerms[topTerms['orcid_id']==orcid2]\n",
    "    m = o1.merge(o2, left_on='term', right_on='term')\n",
    "    return m.iloc[:, [1,2,4]]\n",
    "\n",
    "counter=0\n",
    "oColab=orcidToOrcid[orcidToOrcid['orcid_1']==example_orcid]['orcid_2']\n",
    "for o in oColab:\n",
    "    colabPubs=orcidToPubmed[orcidToPubmed['orcid_id']==o]['pmid']\n",
    "    pubCom=len(set(example_pubs).intersection(set(colabPubs)))\n",
    "    #print(o)\n",
    "    if pubCom==0:\n",
    "        if counter<5:\n",
    "            #get overlapping tf-idf terms\n",
    "            overlapPubs = find_overlapping_tfidf(example_orcid,o)\n",
    "            r = orcidToOrcid[(orcidToOrcid['orcid_1']==example_orcid) & (orcidToOrcid['orcid_2']==o)]\n",
    "            print(o,r['tf-idf'].to_string(header=False))\n",
    "            print(overlapPubs,'\\n')\n",
    "            counter+=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching text to people\n",
    "\n",
    "Using the same tf-idf model, we can try and match any piece of text to the most relevant people."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/55677314/using-sklearn-how-do-i-calculate-the-tf-idf-cosine-similarity-between-documents\n",
    "%store -r tfs\n",
    "corpus_tfidf = tfs\n",
    "%store -r tfidf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " # Comparing text: \n",
      "Diagnosis of Coronary Heart Diseases Using Gene Expression Profiling; Stable Coronary Artery Diseas ...\n",
      "\n",
      " 0.1498055959132002 0000-0002-2515-0852 Chiara Bucciarelliducci\n",
      "myocardial 0.19096778316249413 0.21954367831038224\n",
      "coronary 0.14344988815995552 0.12280020987203692\n",
      "myocardial infarction 0.14301498823094844 0.0809045507024542\n",
      "infarction 0.1376311519301692 0.08278356899359958\n",
      "artery 0.1260582378238358 0.0407330570196053\n",
      "mi 0.08962962902132333 0.03268908257777348\n",
      "disease 0.08950236549400564 0.07128833603762648\n",
      "patients 0.0840322723215066 0.10959539873839973\n",
      "\n",
      " 0.14495921424895872 0000-0003-0924-3247 Tom R Gaunt\n",
      "coronary 0.14344988815995552 0.06361745729766614\n",
      "disease 0.08950236549400564 0.09837724280174748\n",
      "gene 0.06140523925283896 0.09681124182367212\n",
      "genes 0.04930607122405027 0.0931818434280978\n",
      "\n",
      " 0.14106425254783597 0000-0002-1753-3730 Gianni D Angelini\n",
      "myocardial 0.19096778316249413 0.05845243236233413\n",
      "coronary 0.14344988815995552 0.18075640109051788\n",
      "coronary artery 0.13640555940178153 0.16469247334853954\n",
      "artery 0.1260582378238358 0.20013502302147584\n",
      "disease 0.08950236549400564 0.040263270227885534\n",
      "patients 0.0840322723215066 0.15812749668011314\n",
      "ischemia 0.060469669815518816 0.05484969581966551\n",
      "\n",
      " 0.09991253199428538 0000-0003-4638-601X Thomas W Johnson\n",
      "myocardial 0.19096778316249413 0.1209619008070373\n",
      "coronary 0.14344988815995552 0.11189788634413517\n",
      "myocardial infarction 0.14301498823094844 0.115848487922272\n",
      "infarction 0.1376311519301692 0.11445152891408746\n",
      "patients 0.0840322723215066 0.05472077333859413\n",
      "\n",
      " 0.08854748685568083 0000-0001-7240-4563 Richard W Morris\n",
      "coronary 0.14344988815995552 0.12272802981629267\n",
      "mi 0.08962962902132333 0.053985618672753824\n",
      "disease 0.08950236549400564 0.14258630566828698\n",
      "patients 0.0840322723215066 0.05785242883100924\n",
      "\n",
      " # Comparing text: \n",
      "genome wide association gwas\n",
      " ...\n",
      "\n",
      " 0.3257341145113555 0000-0002-4680-3517 M U B St Pourcain\n",
      "gwas 0.3988304963634414 0.057630480441355575\n",
      "association 0.20410747542319485 0.19098945582127533\n",
      "\n",
      " 0.29031595501552243 0000-0002-7141-9189 Nicholas John Timpson\n",
      "gwas 0.3988304963634414 0.04903429704368548\n",
      "association 0.20410747542319485 0.21756422691733515\n",
      "\n",
      " 0.2810249017952208 0000-0003-2514-0889 Lavinia Paternoster\n",
      "gwas 0.3988304963634414 0.061581413303055436\n",
      "association 0.20410747542319485 0.16500871802190314\n",
      "\n",
      " 0.19979060872812424 0000-0003-0920-1055 Gibran Hemani\n",
      "gwas 0.3988304963634414 0.07869423765507806\n",
      "genome 0.27307906008081884 0.03352192842875887\n",
      "association 0.20410747542319485 0.0825115687772545\n",
      "\n",
      " 0.18135603749872564 0000-0002-1407-8314 George Davey Smith\n",
      "association 0.20410747542319485 0.17651501550607201\n",
      "\n",
      " # Comparing text: \n",
      "quantum\n",
      " ...\n",
      "\n",
      " 0.4806612557142124 0000-0001-6184-1813 Jonathan C F Matthews\n",
      "quantum 1.0 0.4573182482253193\n",
      "\n",
      " 0.43927985900065186 0000-0002-1066-3110 Anthony Laing\n",
      "quantum 1.0 0.4169794697317553\n",
      "\n",
      " 0.3884213532226811 0000-0002-3429-4890 Joshua W Silverstone\n",
      "quantum 1.0 0.38446244360735904\n",
      "\n",
      " 0.3477707806390036 0000-0001-9645-0580 Raffaele Santagati\n",
      "quantum 1.0 0.3513039572445175\n",
      "\n",
      " 0.2826585633780625 0000-0003-2944-1988 Karoline Wiesner\n",
      "quantum 1.0 0.2651136169211218\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "#merge pure people data\n",
    "pure_people_orcid = personToOrcid.merge(pureToName, left_on='pure_person_id', right_on='pure_person_id')\n",
    "\n",
    "def compare_text(query):\n",
    "    print('\\n','# Comparing text:',query[:100],'...')\n",
    "    query_tfidf = tfidf.transform([query])\n",
    "    \n",
    "    #get terms for query\n",
    "    feature_names = tfidf.get_feature_names()\n",
    "    res={}\n",
    "    for col in query_tfidf.nonzero()[1]:\n",
    "        res[feature_names[col]]=query_tfidf[0, col]\n",
    "        #reverse sort the results\n",
    "        sorted_res = sorted(res.items(), key=lambda kv: kv[1], reverse=True)\n",
    "    \n",
    "    #compare tfidf\n",
    "    cosineSimilarities = cosine_similarity(query_tfidf, corpus_tfidf).flatten()\n",
    "    cosineData={}\n",
    "    \n",
    "    for c in range(0,len(cosineSimilarities)):\n",
    "        cosineData[c]=cosineSimilarities[c]\n",
    "    sortedCosineData = sorted(cosineData.items(), key=lambda kv: kv[1], reverse=True)\n",
    "    \n",
    "    for s in sortedCosineData[0:5]:\n",
    "        #get pureID \n",
    "        orcid=list(token_dict)[s[0]]\n",
    "        p = pure_people_orcid[personToOrcid['orcid_id']==orcid]\n",
    "        print('\\n',s[1],orcid,p['person_name'].values[0])\n",
    "        \n",
    "        #get overlapping terms\n",
    "        o1=topTerms[topTerms['orcid_id']==orcid]\n",
    "        o2=sorted_res[:100]\n",
    "        #print(o2)\n",
    "        termMatch=[]\n",
    "        #print(o1.term.values)\n",
    "        for i in o2:\n",
    "            #print(i)\n",
    "            if i[0] in o1.term.values:\n",
    "                print(i[0], i[1], o1[o1['term']==i[0]]['tf-idf'].values[0])\n",
    "        \n",
    "        \n",
    "#26930047\n",
    "query1=\"\"\"\n",
    "Diagnosis of Coronary Heart Diseases Using Gene Expression Profiling; Stable Coronary Artery Disease, Cardiac Ischemia with and without Myocardial Necrosis.\n",
    "Cardiovascular disease (including coronary artery disease and myocardial infarction) is one of the leading causes of death in Europe, and is influenced by both \n",
    "environmental and genetic factors. With the recent advances in genomic tools and technologies there is potential to predict and diagnose heart disease using \n",
    "molecular data from analysis of blood cells. We analyzed gene expression data from blood samples taken from normal people (n = 21), non-significant coronary artery \n",
    "disease (n = 93), patients with unstable angina (n = 16), stable coronary artery disease (n = 14) and myocardial infarction (MI; n = 207). We used a feature \n",
    "selection approach to identify a set of gene expression variables which successfully differentiate different cardiovascular diseases. The initial features \n",
    "were discovered by fitting a linear model for each probe set across all arrays of normal individuals and patients with myocardial infarction. Three different \n",
    "feature optimisation algorithms were devised which identified two discriminating sets of genes, one using MI and normal controls (total genes = 6) and another \n",
    "one using MI and unstable angina patients (total genes = 7). In all our classification approaches we used a non-parametric k-nearest neighbour (KNN) classification \n",
    "method (k = 3). The results proved the diagnostic robustness of the final feature sets in discriminating patients with myocardial infarction from healthy controls. \n",
    "Interestingly it also showed efficacy in discriminating myocardial infarction patients from patients with clinical symptoms of cardiac ischemia but no myocardial \n",
    "necrosis or stable coronary artery disease, despite the influence of batch effects and different microarray gene chips and platforms.\n",
    "\"\"\"\n",
    "compare_text(query1)\n",
    "\n",
    "query2=\"\"\"\n",
    "genome wide association gwas\n",
    "\"\"\"\n",
    "compare_text(query2)\n",
    "\n",
    "query3=\"\"\"\n",
    "quantum\n",
    "\"\"\"\n",
    "compare_text(query3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
