{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# To recap\n",
    "\n",
    "### Data so far\n",
    "\n",
    "1. We have created a list of people and their ORCID identifiers - [data/pure_person_to_orcid.txt](data/pure_person_to_orcid.txt)\n",
    "2. We have created a list of ORCID to PubMed identifiers - [data/orcid.tsv](data/orcid.tsv)\n",
    "3. We have created a list of PubMed IDs to PubMed info - [data/pubmed.tsv](data/pubmed.tsv)\n",
    "4. We have created a list of the top 100 terms for each person - [data/orcid-tf-idf.txt](data/orcid-tf-idf.txt)\n",
    "\n",
    "### Questions\n",
    "\n",
    "Now we need to use the common identifiers in these data to answer some questions:\n",
    "\n",
    "1. Can we produce a set of potential collaborators for each person, a collaborator being someone they have significant terms in common with but not previously published with.\n",
    "2. Can we select a set of people that most closely map to a specific piece of text.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pandas dataframes\n",
    "\n",
    "Normally at this point we would use a database, and the phrase `knowledge graph` in the title of the workshop suggests that would be the aim here. However, to keep it all in a single language and framework, and as the data themselves are relatively small, we can use pandas - https://pandas.pydata.org/\n",
    "\n",
    ">pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, following on from the data descriptions above:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         pure_person_id             orcid_id\n",
      "0  4ec18a96-fd2e-4311-a6fa-7ec65696a4e9  0000-0003-3108-5775\n",
      "1  7a04c325-7a39-4560-8308-8cbcaa763747  0000-0002-6772-7111\n",
      "2  b4014828-88e9-4861-ae1d-5c369b6ae35a  0000-0001-7328-4233\n",
      "3  ccedc4d6-6d7e-4a60-8313-f59409ecc6dd  0000-0001-6224-3073\n",
      "4  66e3df0f-6cbe-4cf4-a034-9e1b57684f6b  0000-0003-0300-4990\n",
      "(898, 2)\n",
      "              orcid_id      pmid\n",
      "0  0000-0001-5001-3350  26961927\n",
      "1  0000-0001-5001-3350  30605491\n",
      "2  0000-0001-5008-0705  29118635\n",
      "3  0000-0001-5017-9473  18194108\n",
      "4  0000-0001-5017-9473  18607707\n",
      "(15722, 2)\n",
      "       pmid  year                                              title  \\\n",
      "0  25475436  2015  Sixty-five common genetic variants and predict...   \n",
      "1  25011450  2014  Association between alcohol and cardiovascular...   \n",
      "2  28968714  2018  FATHMM-XF: accurate prediction of pathogenic p...   \n",
      "3  21965548  2012  Four genetic loci influencing electrocardiogra...   \n",
      "4  26930047  2016  Diagnosis of Coronary Heart Diseases Using Gen...   \n",
      "\n",
      "                                            abstract  \n",
      "0  We developed a 65 type 2 diabetes (T2D) varian...  \n",
      "1  To use the rs1229984 variant in the alcohol de...  \n",
      "2  We present FATHMM-XF, a method for predicting ...  \n",
      "3  Presence of left ventricular hypertrophy on an...  \n",
      "4  Cardiovascular disease (including coronary art...  \n",
      "(11160, 4)\n",
      "              orcid_id             term    tf-idf\n",
      "0  0000-0003-0924-3247      methylation  0.201768\n",
      "1  0000-0003-0924-3247         variants  0.139032\n",
      "2  0000-0003-0924-3247             loci  0.138066\n",
      "3  0000-0003-0924-3247          genetic  0.136506\n",
      "4  0000-0003-0924-3247  dna methylation  0.127503\n",
      "(52546, 3)\n",
      "                         pure_person_id         person_name\n",
      "0  f8a64592-741a-4b32-a017-5730498d7462   Stefania Placenti\n",
      "1  5208f654-2f78-4432-b82e-2d68fff58e46   Philip D Harfield\n",
      "2  2eb962d5-5a55-4eca-8136-0dda897f6302     Magnus Feldmann\n",
      "3  eed59a07-5ea6-485b-8e78-5d979a78c553      Steven C Greer\n",
      "4  afffe47a-4eb4-4a84-8393-52701454f60b  Jessica L Fielding\n",
      "(3565, 2)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import config\n",
    "\n",
    "# 1. People and their ORCID identifiers \n",
    "personToOrcid = pd.read_csv(config.demoPureOrcidFile,sep='\\t')\n",
    "print(personToOrcid.head())\n",
    "print(personToOrcid.shape)\n",
    "\n",
    "# 2. ORCID to PubMed identifiers\n",
    "orcidToPubmed = pd.read_csv(config.demoOrcidFile,sep='\\t')\n",
    "print(orcidToPubmed.head())\n",
    "print(orcidToPubmed.shape)\n",
    "\n",
    "# 3. PubMed IDs to PubMed info\n",
    "pubmedToInfo = pd.read_csv(config.demoPubmedFile,sep='\\t')\n",
    "print(pubmedToInfo.head())\n",
    "print(pubmedToInfo.shape)\n",
    "\n",
    "# 4. Top 100 terms for each person\n",
    "topTerms = pd.read_csv(config.demoTfidfFile,sep='\\t')\n",
    "print(topTerms.head())\n",
    "print(topTerms.shape)\n",
    "\n",
    "# 5. PURE to name\n",
    "pureToName = pd.read_csv(config.demoPurePeopleFile,sep='\\t')\n",
    "print(pureToName.head())\n",
    "print(pureToName.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A first pass would be to count the number of matching terms between each person. To start with, let's just compare the first in the list to everyone else:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Comparing 0 4ec18a96-fd2e-4311-a6fa-7ec65696a4e9 0000-0003-3108-5775\n",
      "0000-0001-7086-8882 ['vortex configuration', 'unusually cold temperatures', 'titans', 'results high infrared', 'saturns', 'unexpected rapid', 'southpolar', 'reach stable vortex', 'trace gases relatively', 'years postequinox', 'polar vortex', 'subsidence winter polar', 'substantial nitrogenmethane', 'long', 'postequinox', 'strengthening subsidence', 'winter polar', 'polar vortices following', 'titans winter', 'southpolar subsidence', 'subsidence produced mesospheric', 'vortices following', 'southpolar subsidence winter', 'vortex formation', 'produced trace', 'formation', 'trace gas enrichment', 'titans trace gases', 'temperatures 26 years', 'cooling', 'postequinox long timeframe', 'radiative time constant', 'unexpected rapid mesospheric', 'polar vortex formation', 'trace gases 2012', 'relatively long atmospheric', 'stable vortex', 'saturns largest moon', 'timeframe reach stable', 'titans winter polar', 'titans 2009 northern', 'gases', 'subsidence winter', 'radiative', 'titan', 'stable vortex configuration', 'strong seasonal effects', 'strengthening subsidence produced', 'radiative cooling', 'enrichment', 'vortex saturns largest', 'spring equinox', 'solar heating', 'temperatures 26', 'saturns largest', 'trace gases', 'produced mesospheric', 'extreme', 'mesospheric', 'timeframe reach', 'vortex dramatically', 'postequinox long', 'seasonal effects including', 'vortex dramatically increases', 'polar vortex saturns', 'postequinox cooling', 'vortex', 'subsidence produced', 'vortex saturns', 'produced mesospheric hotspot', 'vortex configuration results', 'years postequinox long', 'radiative time', 'postequinox cooling far', 'vortex formation 20102011', 'winter polar vortices', 'produced trace gases', 'unique titan', 'polar vortices', 'unusually cold', 'rapid mesospheric', 'solar heating moved', 'spring equinox peak', 'vortices following titans', 'winter polar vortex', 'substantial nitrogenmethane atmosphere', 'rapid mesospheric cooling', 'titan substantial', 'seasonal effects', 'titans trace', 'titans 2009', 'polar', 'northern', 'winter polar hotspots', 'trace', 'titan substantial nitrogenmethane', 'time constant winter', 'radiative cooling efficiency', 'subsidence', 'winter']\n"
     ]
    }
   ],
   "source": [
    " #loop through people (limit to top 1)\n",
    "def compare_people(limit):\n",
    "    for i, iData in personToOrcid.head(n=limit).iterrows():\n",
    "        print('### Comparing',i,iData.pure_person_id,iData.orcid_id)\n",
    "        #get all terms and tf-idf values\n",
    "        iTopTerms=topTerms[topTerms['orcid_id']==iData.orcid_id]\n",
    "        #print(iTopTerms.head())\n",
    "        iTerms = iTopTerms['term']\n",
    "\n",
    "        for j in range(i+1,personToOrcid.shape[0]):\n",
    "            jData=personToOrcid.iloc[j]\n",
    "            jTopTerms=topTerms[topTerms['orcid_id']==jData.orcid_id]\n",
    "            #print(jTopTerms['term'].head())\n",
    "            jTerms = jTopTerms['term']\n",
    "            #print(iTerms,jTerms)\n",
    "            com = list(set(iTerms) & set(jTerms))\n",
    "            #only show maches with more than 1 matching term\n",
    "            if len(com) > 1:\n",
    "                print(jData.orcid_id,com)\n",
    "    \n",
    "compare_people(1)        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are few matches with many overlapping terms, except one **0000-0001-7086-8882**. Let's look at the research pages of these two people\n",
    "\n",
    "https://research-information.bristol.ac.uk/en/persons/melody-a-s-sylvestre(81e7f06c-77d8-4020-9608-0b30dd001c43)/publications.html https://research-information.bristol.ac.uk/en/persons/nicholas-a-teanby(4ec18a96-fd2e-4311-a6fa-7ec65696a4e9)/publications.html \n",
    "\n",
    "This suggests they are from similar research areas, and have indeed co-published.   \n",
    "\n",
    "Let's try again, this time with the top 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Comparing 0 4ec18a96-fd2e-4311-a6fa-7ec65696a4e9 0000-0003-3108-5775\n",
      "0000-0001-7086-8882 ['vortex configuration', 'unusually cold temperatures', 'titans', 'results high infrared', 'saturns', 'unexpected rapid', 'southpolar', 'reach stable vortex', 'trace gases relatively', 'years postequinox', 'polar vortex', 'subsidence winter polar', 'substantial nitrogenmethane', 'long', 'postequinox', 'strengthening subsidence', 'winter polar', 'polar vortices following', 'titans winter', 'southpolar subsidence', 'subsidence produced mesospheric', 'vortices following', 'southpolar subsidence winter', 'vortex formation', 'produced trace', 'formation', 'trace gas enrichment', 'titans trace gases', 'temperatures 26 years', 'cooling', 'postequinox long timeframe', 'radiative time constant', 'unexpected rapid mesospheric', 'polar vortex formation', 'trace gases 2012', 'relatively long atmospheric', 'stable vortex', 'saturns largest moon', 'timeframe reach stable', 'titans winter polar', 'titans 2009 northern', 'gases', 'subsidence winter', 'radiative', 'titan', 'stable vortex configuration', 'strong seasonal effects', 'strengthening subsidence produced', 'radiative cooling', 'enrichment', 'vortex saturns largest', 'spring equinox', 'solar heating', 'temperatures 26', 'saturns largest', 'trace gases', 'produced mesospheric', 'extreme', 'mesospheric', 'timeframe reach', 'vortex dramatically', 'postequinox long', 'seasonal effects including', 'vortex dramatically increases', 'polar vortex saturns', 'postequinox cooling', 'vortex', 'subsidence produced', 'vortex saturns', 'produced mesospheric hotspot', 'vortex configuration results', 'years postequinox long', 'radiative time', 'postequinox cooling far', 'vortex formation 20102011', 'winter polar vortices', 'produced trace gases', 'unique titan', 'polar vortices', 'unusually cold', 'rapid mesospheric', 'solar heating moved', 'spring equinox peak', 'vortices following titans', 'winter polar vortex', 'substantial nitrogenmethane atmosphere', 'rapid mesospheric cooling', 'titan substantial', 'seasonal effects', 'titans trace', 'titans 2009', 'polar', 'northern', 'winter polar hotspots', 'trace', 'titan substantial nitrogenmethane', 'time constant winter', 'radiative cooling efficiency', 'subsidence', 'winter']\n",
      "### Comparing 1 7a04c325-7a39-4560-8308-8cbcaa763747 0000-0002-6772-7111\n",
      "### Comparing 2 b4014828-88e9-4861-ae1d-5c369b6ae35a 0000-0001-7328-4233\n",
      "0000-0001-6224-3073 ['functional', 'molecular', 'using']\n",
      "0000-0001-7382-2855 ['database', 'molecular', 'using', 'analysis', 'different', 'diversity', 'data']\n",
      "0000-0003-2881-4180 ['analysis', 'using', 'data', 'cancer']\n",
      "0000-0002-1753-3730 ['cell', 'angiogenesis']\n",
      "0000-0002-1040-7635 ['data', 'cancer']\n",
      "0000-0001-8496-6053 ['analysis', 'using', 'data', 'cancer']\n",
      "0000-0002-9848-1409 ['stem cell', 'stem', 'cell']\n",
      "0000-0002-2618-750X ['taxa', 'functional', 'using', 'analysis', 'species', 'data']\n",
      "0000-0002-5555-433X ['breast cancer', 'cancer', 'breast']\n",
      "0000-0002-2497-1026 ['using', 'cell', 'cellular']\n",
      "0000-0001-7206-3674 ['phenotype', 'data']\n",
      "0000-0001-5990-9161 ['sequence', 'data']\n",
      "0000-0002-3510-4155 ['molecular', 'using']\n",
      "0000-0002-4323-1824 ['diversity', 'molecular', 'clades', 'species']\n",
      "0000-0002-7961-8027 ['heritability', 'development']\n",
      "0000-0003-1561-9933 ['cell', 'ld', 'cellular']\n",
      "0000-0002-5311-6213 ['using', 'analysis', 'genomes', 'diversity', 'sequence', 'data', 'genome']\n",
      "0000-0002-9732-5738 ['nematode', 'using', 'parasitic', 'species', 'nematode species', 'diversity', 'nematodes', 'data']\n",
      "0000-0003-0574-5071 ['analysis', 'using', 'data']\n",
      "0000-0002-1035-9020 ['molecular', 'different']\n",
      "0000-0002-7404-9128 ['analysis', 'data']\n",
      "0000-0001-8186-5708 ['cell', 'cancer']\n",
      "0000-0002-3097-3765 ['genes', 'stem', 'cell']\n",
      "0000-0002-4032-9500 ['potential', 'using', 'data', 'ld']\n",
      "0000-0003-3909-6698 ['mechanisms', 'stem']\n",
      "0000-0002-0125-4841 ['cancer', 'cell', 'development']\n",
      "0000-0001-6012-6527 ['analysis', 'different']\n",
      "0000-0002-4049-993X ['using', 'data']\n",
      "0000-0003-4611-8795 ['cell', 'genes', 'cellular', 'analysis', 'sequence', 'data', 'genome']\n",
      "0000-0001-7793-7326 ['mouth', 'mouth ulcers', 'heritability', 'data', 'ulcers']\n",
      "0000-0002-8608-289X ['analysis', 'heritability', 'genes', 'development']\n",
      "0000-0002-5943-3981 ['cell', 'molecular', 'cellular', 'resource', 'data']\n",
      "0000-0003-1217-4852 ['genes', 'cell', 'development']\n",
      "0000-0003-3116-7463 ['taxa', 'mirna', 'molecular', 'using', 'clades', 'development', 'species', 'diversity', 'data', 'genome']\n",
      "0000-0002-1113-7761 ['stem', 'cell', 'expressed', 'sequence', 'transcripts']\n",
      "0000-0001-7808-6121 ['functional', 'stem cell', 'stem', 'cell']\n",
      "0000-0001-5673-9371 ['phenotype', 'mechanisms', 'cell', 'cellular']\n",
      "0000-0003-2906-4035 ['analysis', 'mechanisms']\n",
      "0000-0002-0394-3202 ['functional', 'potential', 'sequences', 'analysis', 'sequence']\n",
      "0000-0003-2946-0353 ['genes', 'sequences', 'cell', 'expressed']\n",
      "0000-0002-5312-0378 ['molecular', 'using']\n",
      "0000-0002-6177-577X ['genes', 'mechanisms', 'species', 'genome']\n",
      "0000-0002-2393-4522 ['stem cell', 'stem', 'cell']\n",
      "0000-0002-2183-8166 ['cancer', 'breast']\n",
      "0000-0002-1573-7908 ['using', 'species']\n",
      "0000-0003-0933-646X ['cancer', 'breast', 'analysis', 'breast cancer', 'data', 'genome']\n",
      "0000-0003-3631-627X ['using', 'data', 'cancer']\n",
      "0000-0003-3565-0479 ['using', 'cell']\n",
      "0000-0002-6977-312X ['breast cancer', 'cancer', 'breast']\n",
      "0000-0002-1407-8314 ['genes', 'using', 'analysis', 'data', 'cancer']\n",
      "0000-0001-5255-8506 ['potential', 'data']\n",
      "0000-0003-3979-7857 ['potential', 'molecular']\n",
      "0000-0001-8202-1144 ['using', 'data']\n",
      "0000-0003-3113-0345 ['molecular', 'data']\n",
      "0000-0002-9528-1495 ['analysis', 'species', 'cellular']\n",
      "0000-0003-1454-8440 ['genes', 'sequences', 'sequence', 'genome']\n",
      "0000-0002-8152-6017 ['using', 'data', 'different']\n",
      "0000-0003-3719-6361 ['cancer', 'different']\n",
      "0000-0002-3853-883X ['potential', 'using', 'cancer', 'development']\n",
      "0000-0002-0935-2200 ['analysis', 'genes', 'data', 'genome']\n",
      "0000-0002-7400-005X ['using', 'species']\n",
      "0000-0003-1562-891X ['cancer cell', 'cell', 'development', 'breast', 'breast cancer', 'cancer']\n",
      "0000-0002-2861-0578 ['database', 'cancer', 'data', 'different']\n",
      "0000-0003-2052-4840 ['genes', 'using', 'cancer', 'development', 'analysis', 'data', 'genome']\n",
      "0000-0002-4701-8713 ['cell', 'development']\n",
      "0000-0003-3894-0623 ['cell', 'angiogenesis']\n",
      "0000-0002-7708-3110 ['annotated', 'data']\n",
      "0000-0002-2115-3503 ['molecular', 'cellular', 'cell', 'genome']\n",
      "0000-0002-3897-5861 ['phenotype', 'cell', 'stem', 'stem cell']\n",
      "0000-0002-7049-3037 ['analysis', 'data', 'cancer']\n",
      "0000-0003-3049-0939 ['using', 'cell']\n",
      "0000-0003-3756-040X ['mouth', 'mouth ulcers', 'data', 'ulcers', 'cancer']\n",
      "0000-0002-8911-2492 ['heritability', 'data']\n",
      "0000-0002-8917-7384 ['cancer cell', 'cell', 'tumor', 'cellular', 'breast cancer', 'cancer']\n",
      "0000-0001-6648-3007 ['analysis', 'using', 'data']\n",
      "0000-0001-8923-0117 ['analysis', 'using', 'cancer']\n",
      "0000-0002-4680-3517 ['genes', 'using', 'development', 'analysis', 'data']\n",
      "0000-0002-2870-1832 ['data', 'development']\n",
      "0000-0001-8091-6534 ['molecular', 'cell']\n",
      "0000-0001-5297-3240 ['functional', 'cell']\n",
      "0000-0001-5747-2699 ['data', 'cancer']\n",
      "0000-0002-2501-6602 ['molecular', 'sequence']\n",
      "0000-0003-0848-935X ['genes', 'mechanisms']\n",
      "0000-0003-1072-0223 ['functional', 'genes', 'freeliving', 'molecular', 'cellular', 'analysis', 'genomes', 'diversity', 'sequence', 'data', 'genome']\n",
      "0000-0003-0924-3247 ['functional', 'potential', 'genes', 'using', 'cancer', 'analysis', 'score', 'data', 'genome']\n",
      "0000-0002-4540-6683 ['species', 'different']\n",
      "0000-0001-7652-2030 ['species', 'development']\n",
      "0000-0002-4612-7141 ['sequences', 'sequence', 'different']\n",
      "0000-0002-5904-8397 ['analysis', 'using', 'data']\n",
      "0000-0003-2601-7575 ['breast cancer', 'cell', 'breast']\n",
      "0000-0003-4395-9396 ['cell', 'expressed']\n",
      "0000-0002-2601-9258 ['cancer', 'data', 'development']\n",
      "0000-0002-2715-9930 ['genes', 'mechanisms', 'development', 'data', 'cancer']\n",
      "0000-0001-6566-9205 ['mechanisms', 'stem']\n",
      "0000-0002-3548-6245 ['analysis', 'cell']\n",
      "0000-0003-2656-4034 ['microrna', 'mirna', 'stem', 'mechanisms', 'micrornas']\n",
      "0000-0003-0920-1055 ['genes', 'using', 'ld', 'analysis', 'heritability', 'data', 'genome']\n",
      "0000-0002-7141-9189 ['analysis', 'genes', 'using', 'data']\n",
      "0000-0003-3405-4843 ['sequences', 'genes', 'using', 'analysis', 'species']\n",
      "0000-0003-4416-4560 ['molecular', 'species']\n",
      "0000-0001-8539-8959 ['mirna', 'cell']\n",
      "0000-0003-0858-6088 ['potential', 'cell']\n",
      "0000-0002-7335-9001 ['genes', 'sequence']\n",
      "0000-0001-6849-8747 ['diversity', 'resource', 'species']\n",
      "0000-0002-8773-2743 ['molecular', 'using', 'analysis', 'species', 'data', 'genome']\n",
      "0000-0001-9595-3338 ['molecular', 'different']\n",
      "0000-0002-9759-5156 ['functional', 'phenotype', 'stem', 'cell']\n",
      "0000-0001-6293-6662 ['analysis', 'sequence', 'species']\n",
      "0000-0001-9426-686X ['breast cancer', 'cancer', 'cell', 'breast']\n",
      "0000-0002-3354-3330 ['development', 'cancer', 'data', 'breast']\n",
      "0000-0002-9133-2021 ['genes', 'data', 'development']\n",
      "0000-0001-8548-1090 ['using', 'molecular', 'sequence']\n",
      "0000-0002-4217-3060 ['diversity', 'data', 'different']\n",
      "0000-0003-1486-745X ['analysis', 'data']\n",
      "0000-0003-2826-3307 ['breast', 'using', 'data', 'development']\n",
      "0000-0002-6793-2262 ['analysis', 'using', 'data']\n",
      "0000-0002-4935-6724 ['genes', 'cell', 'development']\n",
      "0000-0002-1010-8926 ['using', 'data', 'score', 'cancer']\n",
      "0000-0001-6678-8839 ['molecular', 'data']\n",
      "0000-0003-0609-475X ['analysis', 'functional']\n",
      "0000-0003-4690-8826 ['functional', 'cell', 'molecular', 'analysis', 'different']\n",
      "0000-0002-6009-7137 ['cell', 'stem', 'tumor', 'cancer']\n",
      "0000-0003-3589-8335 ['functional', 'molecular', 'mechanisms', 'cell']\n",
      "0000-0003-2088-8310 ['analysis', 'using', 'cell']\n",
      "0000-0002-6828-5888 ['stem', 'cell', 'cellular']\n",
      "0000-0002-3584-9616 ['taxa', 'molecular', 'stem']\n",
      "0000-0002-5300-3331 ['functional', 'potential', 'stem', 'cell', 'development', 'stem cell']\n",
      "0000-0003-2198-3205 ['analysis', 'using', 'data', 'different']\n",
      "0000-0002-3922-0188 ['analysis', 'genes']\n",
      "0000-0002-0950-8148 ['genes', 'phenotype', 'genome']\n",
      "0000-0003-1702-786X ['using', 'genome']\n",
      "0000-0001-5551-9609 ['molecular', 'using']\n",
      "0000-0002-9644-8998 ['genes', 'diversity', 'species', 'different']\n",
      "0000-0002-3537-5166 ['using', 'analysis', 'different', 'resource', 'data']\n",
      "0000-0003-2070-6679 ['sequences', 'using', 'different']\n",
      "0000-0002-3208-0389 ['genes', 'heritability']\n",
      "0000-0002-1904-3735 ['sequences', 'genes', 'molecular', 'analysis', 'species']\n",
      "0000-0002-0546-2381 ['using', 'data']\n",
      "0000-0003-2514-0889 ['genes', 'data']\n",
      "0000-0002-6391-6601 ['functional', 'molecular']\n",
      "0000-0002-4701-793X ['mechanisms', 'species']\n",
      "0000-0002-7992-7719 ['using', 'data', 'cancer']\n",
      "0000-0001-5677-5401 ['functional', 'using', 'species', 'diversity', 'parasitic', 'data']\n",
      "0000-0001-8595-5153 ['stem', 'cell']\n",
      "0000-0001-7240-4563 ['analysis', 'using', 'data', 'score']\n"
     ]
    }
   ],
   "source": [
    "compare_people(3)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second person matched no-one, the third however matched quite a few. Let's modify the function to sort by number:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Comparing 0 4ec18a96-fd2e-4311-a6fa-7ec65696a4e9 0000-0003-3108-5775\n",
      "('0000-0001-7086-8882', ['vortex configuration', 'unusually cold temperatures', 'titans', 'results high infrared', 'saturns', 'unexpected rapid', 'southpolar', 'reach stable vortex', 'trace gases relatively', 'years postequinox', 'polar vortex', 'subsidence winter polar', 'substantial nitrogenmethane', 'long', 'postequinox', 'strengthening subsidence', 'winter polar', 'polar vortices following', 'titans winter', 'southpolar subsidence', 'subsidence produced mesospheric', 'vortices following', 'southpolar subsidence winter', 'vortex formation', 'produced trace', 'formation', 'trace gas enrichment', 'titans trace gases', 'temperatures 26 years', 'cooling', 'postequinox long timeframe', 'radiative time constant', 'unexpected rapid mesospheric', 'polar vortex formation', 'trace gases 2012', 'relatively long atmospheric', 'stable vortex', 'saturns largest moon', 'timeframe reach stable', 'titans winter polar', 'titans 2009 northern', 'gases', 'subsidence winter', 'radiative', 'titan', 'stable vortex configuration', 'strong seasonal effects', 'strengthening subsidence produced', 'radiative cooling', 'enrichment', 'vortex saturns largest', 'spring equinox', 'solar heating', 'temperatures 26', 'saturns largest', 'trace gases', 'produced mesospheric', 'extreme', 'mesospheric', 'timeframe reach', 'vortex dramatically', 'postequinox long', 'seasonal effects including', 'vortex dramatically increases', 'polar vortex saturns', 'postequinox cooling', 'vortex', 'subsidence produced', 'vortex saturns', 'produced mesospheric hotspot', 'vortex configuration results', 'years postequinox long', 'radiative time', 'postequinox cooling far', 'vortex formation 20102011', 'winter polar vortices', 'produced trace gases', 'unique titan', 'polar vortices', 'unusually cold', 'rapid mesospheric', 'solar heating moved', 'spring equinox peak', 'vortices following titans', 'winter polar vortex', 'substantial nitrogenmethane atmosphere', 'rapid mesospheric cooling', 'titan substantial', 'seasonal effects', 'titans trace', 'titans 2009', 'polar', 'northern', 'winter polar hotspots', 'trace', 'titan substantial nitrogenmethane', 'time constant winter', 'radiative cooling efficiency', 'subsidence', 'winter'])\n",
      "### Comparing 1 7a04c325-7a39-4560-8308-8cbcaa763747 0000-0002-6772-7111\n",
      "### Comparing 2 b4014828-88e9-4861-ae1d-5c369b6ae35a 0000-0001-7328-4233\n",
      "('0000-0003-1072-0223', ['functional', 'genes', 'freeliving', 'molecular', 'cellular', 'analysis', 'genomes', 'diversity', 'sequence', 'data', 'genome'])\n",
      "('0000-0003-3116-7463', ['taxa', 'mirna', 'molecular', 'using', 'clades', 'development', 'species', 'diversity', 'data', 'genome'])\n",
      "('0000-0003-0924-3247', ['functional', 'potential', 'genes', 'using', 'cancer', 'analysis', 'score', 'data', 'genome'])\n",
      "('0000-0002-9732-5738', ['nematode', 'using', 'parasitic', 'species', 'nematode species', 'diversity', 'nematodes', 'data'])\n",
      "('0000-0001-7382-2855', ['database', 'molecular', 'using', 'analysis', 'different', 'diversity', 'data'])\n",
      "('0000-0002-5311-6213', ['using', 'analysis', 'genomes', 'diversity', 'sequence', 'data', 'genome'])\n",
      "('0000-0003-4611-8795', ['cell', 'genes', 'cellular', 'analysis', 'sequence', 'data', 'genome'])\n",
      "('0000-0003-2052-4840', ['genes', 'using', 'cancer', 'development', 'analysis', 'data', 'genome'])\n",
      "('0000-0003-0920-1055', ['genes', 'using', 'ld', 'analysis', 'heritability', 'data', 'genome'])\n",
      "('0000-0002-2618-750X', ['taxa', 'functional', 'using', 'analysis', 'species', 'data'])\n"
     ]
    }
   ],
   "source": [
    "def compare_people(limit):\n",
    "    for i, iData in personToOrcid.head(n=limit).iterrows():\n",
    "        print('### Comparing',i,iData.pure_person_id,iData.orcid_id)\n",
    "        #get all terms and tf-idf values\n",
    "        iTopTerms=topTerms[topTerms['orcid_id']==iData.orcid_id]\n",
    "        iTerms = iTopTerms['term']\n",
    "        jComp={}\n",
    "        for j in range(i+1,personToOrcid.shape[0]):\n",
    "            jData=personToOrcid.iloc[j]\n",
    "            jTopTerms=topTerms[topTerms['orcid_id']==jData.orcid_id]\n",
    "            jTerms = jTopTerms['term']\n",
    "            com = list(set(iTerms) & set(jTerms))\n",
    "            #only show maches with more than 1 matching term\n",
    "            if len(com) > 1:\n",
    "                jComp[jData.orcid_id]=com\n",
    "        #create sorted dictionary using number of items\n",
    "        jComp = sorted(jComp.items(), key=lambda kv: len(kv[1]), reverse=True)\n",
    "        for p in jComp[0:10]:\n",
    "            print(p)\n",
    "            \n",
    "compare_people(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This only shows us terms that are in common. It could be possible to have lots of terms in common, but these terms are not the 'most' representative of either person, or indeed in order of importance per person. It would be better to compare all terms, and their tf-idf values simultaneously.\n",
    "\n",
    "Back in the TF-IDF section (http://localhost:8888/lab#TF-IDF-using-sklearn) we created a similarity matrix from the TF-IDF model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0000-0001-5001-3350\n",
      "\t 0000-0001-5001-3350 0.9999999999999963\n",
      "\t 0000-0002-1407-8314 0.15780716812089102\n",
      "\t 0000-0003-2052-4840 0.12414499704387866\n",
      "\t 0000-0003-0920-1055 0.10864635054285894\n",
      "\t 0000-0002-7141-9189 0.0969238856366499\n",
      "0000-0001-5008-0705\n",
      "\t 0000-0001-5008-0705 0.9999999999999997\n",
      "\t 0000-0003-4416-4560 0.019256884487874857\n",
      "\t 0000-0002-2960-759X 0.019226598250719203\n",
      "\t 0000-0002-1904-3735 0.01522090025025024\n",
      "\t 0000-0003-3405-4843 0.01378333900366982\n",
      "0000-0001-5017-9473\n",
      "\t 0000-0001-5017-9473 1.00000000000002\n",
      "\t 0000-0002-8337-626X 0.11511608769135045\n",
      "\t 0000-0001-5396-3108 0.10194269665348185\n",
      "\t 0000-0002-2064-3359 0.10060758508049958\n",
      "\t 0000-0002-7706-8105 0.09074606109190216\n"
     ]
    }
   ],
   "source": [
    "#get similarity matrix for all people\n",
    "\n",
    "matrixCom = {}\n",
    "\n",
    "#get tfidf matrix and orcidText dictionary\n",
    "%store -r matrix\n",
    "%store -r token_dict\n",
    "\n",
    "for i in range(0,len(matrix)):\n",
    "    iOrcid=list(token_dict)[i]\n",
    "    matrixCom[iOrcid]={}\n",
    "    for j in range(0,len(matrix)):\n",
    "        jOrcid=list(token_dict)[j]\n",
    "        matrixCom[iOrcid][jOrcid]=matrix[i][j]\n",
    "\n",
    "o = open(config.orcidToOrcid,'w')\n",
    "o.write('orcid_1\\torcid_2\\ttf-idf\\n')\n",
    "counter=0\n",
    "for m in matrixCom:\n",
    "    sorted_res = sorted(matrixCom[m].items(), key=lambda kv: kv[1], reverse=True)\n",
    "    if counter<3:\n",
    "        print(m)\n",
    "        for s in sorted_res[0:5]:\n",
    "            print('\\t',s[0],s[1])\n",
    "        counter+=1\n",
    "    for s in sorted_res:\n",
    "        o.write(m+'\\t'+s[0]+'\\t'+str(s[1])+'\\n')\n",
    "o.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Collaboration recommendation engine\n",
    "\n",
    "This similarity data depicts the similarity between each person's publication text based on tf-idf. Often, similarities arise due to co-publication, and perhaps a more informative recommender would be to idenfity cases where people have signficant overlap in their publication text, but have never previously co-published. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(279841, 3)\n",
      "               orcid_1              orcid_2    tf-idf\n",
      "0  0000-0001-5001-3350  0000-0001-5001-3350  1.000000\n",
      "1  0000-0001-5001-3350  0000-0002-1407-8314  0.157807\n",
      "2  0000-0001-5001-3350  0000-0003-2052-4840  0.124145\n",
      "3  0000-0001-5001-3350  0000-0003-0920-1055  0.108646\n",
      "4  0000-0001-5001-3350  0000-0002-7141-9189  0.096924\n"
     ]
    }
   ],
   "source": [
    "#load the new data into a dataframe\n",
    "\n",
    "orcidToOrcid = pd.read_csv(config.demoOrcidToOrcid,sep='\\t')\n",
    "print(orcidToOrcid.shape)\n",
    "print(orcidToOrcid.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0000-0002-1407-8314 33334    0.10258\n",
      "       term  tf-idf_x  tf-idf_y\n",
      "0    cancer  0.085914  0.079305\n",
      "1     genes  0.061123  0.050338\n",
      "2      data  0.054236  0.064645\n",
      "3     using  0.035485  0.060613\n",
      "4  analysis  0.032270  0.051746 \n",
      "\n",
      "0000-0003-1562-891X 33335    0.098\n",
      "            term  tf-idf_x  tf-idf_y\n",
      "0           cell  0.091965  0.151794\n",
      "1         cancer  0.085914  0.264330\n",
      "2         breast  0.059255  0.247853\n",
      "3  breast cancer  0.038263  0.198393\n",
      "4    cancer cell  0.032882  0.042547\n",
      "5    development  0.031938  0.032471 \n",
      "\n",
      "0000-0001-8186-5708 33340    0.074197\n",
      "     term  tf-idf_x  tf-idf_y\n",
      "0    cell  0.091965  0.274139\n",
      "1  cancer  0.085914  0.062854 \n",
      "\n",
      "0000-0002-2715-9930 33341    0.070891\n",
      "          term  tf-idf_x  tf-idf_y\n",
      "0       cancer  0.085914  0.042130\n",
      "1        genes  0.061123  0.115058\n",
      "2         data  0.054236  0.045467\n",
      "3   mechanisms  0.040592  0.030163\n",
      "4  development  0.031938  0.035598 \n",
      "\n",
      "0000-0002-6793-2262 33342    0.070718\n",
      "       term  tf-idf_x  tf-idf_y\n",
      "0      data  0.054236  0.051510\n",
      "1     using  0.035485  0.044307\n",
      "2  analysis  0.032270  0.040558 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#filter on co-publishing\n",
    "example_orcid='0000-0001-7328-4233'\n",
    "#example_orcid='0000-0003-0924-3247'\n",
    "example_pubs=orcidToPubmed[orcidToPubmed['orcid_id']==example_orcid]['pmid']\n",
    "\n",
    "def find_overlapping_tfidf(orcid1,orcid2):\n",
    "    o1=topTerms[topTerms['orcid_id']==orcid1]\n",
    "    o2=topTerms[topTerms['orcid_id']==orcid2]\n",
    "    m = o1.merge(o2, left_on='term', right_on='term')\n",
    "    return m.iloc[:, [1,2,4]]\n",
    "\n",
    "counter=0\n",
    "oColab=orcidToOrcid[orcidToOrcid['orcid_1']==example_orcid]['orcid_2']\n",
    "for o in oColab:\n",
    "    colabPubs=orcidToPubmed[orcidToPubmed['orcid_id']==o]['pmid']\n",
    "    pubCom=len(set(example_pubs).intersection(set(colabPubs)))\n",
    "    #print(o)\n",
    "    if pubCom==0:\n",
    "        if counter<5:\n",
    "            #get overlapping tf-idf terms\n",
    "            overlapPubs = find_overlapping_tfidf(example_orcid,o)\n",
    "            r = orcidToOrcid[(orcidToOrcid['orcid_1']==example_orcid) & (orcidToOrcid['orcid_2']==o)]\n",
    "            print(o,r['tf-idf'].to_string(header=False))\n",
    "            print(overlapPubs,'\\n')\n",
    "            counter+=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching text to people\n",
    "\n",
    "Using the same tf-idf model, we can try and match any piece of text to the most relevant people."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/55677314/using-sklearn-how-do-i-calculate-the-tf-idf-cosine-similarity-between-documents\n",
    "%store -r tfs\n",
    "corpus_tfidf = tfs\n",
    "%store -r tfidf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " # Comparing text: \n",
      "Diagnosis of Coronary Heart Diseases Using Gene Expression Profiling; Stable Coronary Artery Diseas ...\n",
      "\n",
      " 0.1498063952666541 0000-0002-2515-0852 Chiara Bucciarelliducci\n",
      "myocardial 0.19095723139805437 0.21954367831038224\n",
      "coronary 0.1434419619568876 0.12280020987203692\n",
      "myocardial infarction 0.1430070860579108 0.0809045507024542\n",
      "infarction 0.13762354723648376 0.08278356899359958\n",
      "artery 0.12605127258179769 0.0407330570196053\n",
      "mi 0.08962467661146345 0.03268908257777348\n",
      "disease 0.08949742011598505 0.07128833603762648\n",
      "patients 0.08402762918888922 0.10959539873839973\n",
      "\n",
      " 0.14473219216074815 0000-0003-0924-3247 Tom R Gaunt\n",
      "coronary 0.1434419619568876 0.06361745729766614\n",
      "disease 0.08949742011598505 0.09837724280174748\n",
      "gene 0.061401846357926365 0.09681124182367212\n",
      "genes 0.049303346858503426 0.0931818434280978\n",
      "\n",
      " 0.14106423986554967 0000-0002-1753-3730 Gianni D Angelini\n",
      "myocardial 0.19095723139805437 0.05845243236233413\n",
      "coronary 0.1434419619568876 0.18075640109051788\n",
      "coronary artery 0.13639802242718171 0.16469247334853954\n",
      "artery 0.12605127258179769 0.20013502302147584\n",
      "disease 0.08949742011598505 0.040263270227885534\n",
      "patients 0.08402762918888922 0.15812749668011314\n",
      "ischemia 0.06046632861470957 0.05484969581966551\n",
      "\n",
      " 0.09992455637447066 0000-0003-4638-601X Thomas W Johnson\n",
      "myocardial 0.19095723139805437 0.1209619008070373\n",
      "coronary 0.1434419619568876 0.11189788634413517\n",
      "myocardial infarction 0.1430070860579108 0.115848487922272\n",
      "infarction 0.13762354723648376 0.11445152891408746\n",
      "patients 0.08402762918888922 0.05472077333859413\n",
      "\n",
      " 0.08853169646516287 0000-0001-7240-4563 Richard W Morris\n",
      "coronary 0.1434419619568876 0.12272802981629267\n",
      "mi 0.08962467661146345 0.053985618672753824\n",
      "disease 0.08949742011598505 0.14258630566828698\n",
      "patients 0.08402762918888922 0.05785242883100924\n",
      "\n",
      " # Comparing text: \n",
      "genome wide association gwas\n",
      " ...\n",
      "\n",
      " 0.32574371740903496 0000-0002-4680-3517 M U B St Pourcain\n",
      "gwas 0.3988304963634414 0.057630480441355575\n",
      "association 0.20410747542319485 0.19098945582127533\n",
      "\n",
      " 0.2903208869809313 0000-0002-7141-9189 Nicholas John Timpson\n",
      "gwas 0.3988304963634414 0.04903429704368548\n",
      "association 0.20410747542319485 0.21756422691733515\n",
      "\n",
      " 0.2810170481824614 0000-0003-2514-0889 Lavinia Paternoster\n",
      "gwas 0.3988304963634414 0.061581413303055436\n",
      "association 0.20410747542319485 0.16500871802190314\n",
      "\n",
      " 0.19976744034069038 0000-0003-0920-1055 Gibran Hemani\n",
      "gwas 0.3988304963634414 0.07869423765507806\n",
      "genome 0.27307906008081884 0.03352192842875887\n",
      "association 0.20410747542319485 0.0825115687772545\n",
      "\n",
      " 0.1813558169199878 0000-0002-1407-8314 George Davey Smith\n",
      "association 0.20410747542319485 0.17651501550607201\n",
      "\n",
      " # Comparing text: \n",
      "quantum mechanics\n",
      " ...\n",
      "\n",
      " 0.28945476707203033 0000-0001-6184-1813 Jonathan C F Matthews\n",
      "quantum mechanics 0.6680585375945389 0.04294802306764767\n",
      "mechanics 0.5305678604555767 0.031977101176721576\n",
      "quantum 0.5217236201272065 0.4573182482253193\n",
      "\n",
      " 0.24824713212689506 0000-0002-1066-3110 Anthony Laing\n",
      "quantum 0.5217236201272065 0.4169794697317553\n",
      "\n",
      " 0.22624172866293998 0000-0002-3429-4890 Joshua W Silverstone\n",
      "quantum 0.5217236201272065 0.38446244360735904\n",
      "\n",
      " 0.18502285880883648 0000-0003-2944-1988 Karoline Wiesner\n",
      "quantum 0.5217236201272065 0.2651136169211218\n",
      "\n",
      " 0.18145107274095362 0000-0001-9645-0580 Raffaele Santagati\n",
      "quantum 0.5217236201272065 0.3513039572445175\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "#merge pure people data\n",
    "pure_people_orcid = personToOrcid.merge(pureToName, left_on='pure_person_id', right_on='pure_person_id')\n",
    "\n",
    "def compare_text(query):\n",
    "    print('\\n','# Comparing text:',query[:100],'...')\n",
    "    query_tfidf = tfidf.transform([query])\n",
    "    \n",
    "    #get terms for query\n",
    "    feature_names = tfidf.get_feature_names()\n",
    "    res={}\n",
    "    for col in query_tfidf.nonzero()[1]:\n",
    "        res[feature_names[col]]=query_tfidf[0, col]\n",
    "        #reverse sort the results\n",
    "        sorted_res = sorted(res.items(), key=lambda kv: kv[1], reverse=True)\n",
    "    \n",
    "    #compare tfidf\n",
    "    cosineSimilarities = cosine_similarity(query_tfidf, corpus_tfidf).flatten()\n",
    "    cosineData={}\n",
    "    \n",
    "    for c in range(0,len(cosineSimilarities)):\n",
    "        cosineData[c]=cosineSimilarities[c]\n",
    "    sortedCosineData = sorted(cosineData.items(), key=lambda kv: kv[1], reverse=True)\n",
    "    \n",
    "    for s in sortedCosineData[0:5]:\n",
    "        #get pureID \n",
    "        orcid=list(token_dict)[s[0]]\n",
    "        p = pure_people_orcid[personToOrcid['orcid_id']==orcid]\n",
    "        print('\\n',s[1],orcid,p['person_name'].values[0])\n",
    "        \n",
    "        #get overlapping terms\n",
    "        o1=topTerms[topTerms['orcid_id']==orcid]\n",
    "        o2=sorted_res[:100]\n",
    "        #print(o2)\n",
    "        termMatch=[]\n",
    "        #print(o1.term.values)\n",
    "        for i in o2:\n",
    "            #print(i)\n",
    "            if i[0] in o1.term.values:\n",
    "                print(i[0], i[1], o1[o1['term']==i[0]]['tf-idf'].values[0])\n",
    "        \n",
    "        \n",
    "#26930047\n",
    "query1=\"\"\"\n",
    "Diagnosis of Coronary Heart Diseases Using Gene Expression Profiling; Stable Coronary Artery Disease, Cardiac Ischemia with and without Myocardial Necrosis.\n",
    "Cardiovascular disease (including coronary artery disease and myocardial infarction) is one of the leading causes of death in Europe, and is influenced by both \n",
    "environmental and genetic factors. With the recent advances in genomic tools and technologies there is potential to predict and diagnose heart disease using \n",
    "molecular data from analysis of blood cells. We analyzed gene expression data from blood samples taken from normal people (n = 21), non-significant coronary artery \n",
    "disease (n = 93), patients with unstable angina (n = 16), stable coronary artery disease (n = 14) and myocardial infarction (MI; n = 207). We used a feature \n",
    "selection approach to identify a set of gene expression variables which successfully differentiate different cardiovascular diseases. The initial features \n",
    "were discovered by fitting a linear model for each probe set across all arrays of normal individuals and patients with myocardial infarction. Three different \n",
    "feature optimisation algorithms were devised which identified two discriminating sets of genes, one using MI and normal controls (total genes = 6) and another \n",
    "one using MI and unstable angina patients (total genes = 7). In all our classification approaches we used a non-parametric k-nearest neighbour (KNN) classification \n",
    "method (k = 3). The results proved the diagnostic robustness of the final feature sets in discriminating patients with myocardial infarction from healthy controls. \n",
    "Interestingly it also showed efficacy in discriminating myocardial infarction patients from patients with clinical symptoms of cardiac ischemia but no myocardial \n",
    "necrosis or stable coronary artery disease, despite the influence of batch effects and different microarray gene chips and platforms.\n",
    "\"\"\"\n",
    "compare_text(query1)\n",
    "\n",
    "query2=\"\"\"\n",
    "genome wide association gwas\n",
    "\"\"\"\n",
    "compare_text(query2)\n",
    "\n",
    "query3=\"\"\"\n",
    "quantum mechanics\n",
    "\"\"\"\n",
    "compare_text(query3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
